tests/test_interpolation.py

```diff
@@ -753,3 +755,244 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
+            7,
+            IntegerNode,
+            id="intermediate_type_mismatch_ok",
+        ),
+        pytest.param(
+            # This example relies on the automatic casting of a string to int when
+            # assigned to an IntegerNode.
+            User(name="Bond", age=SI("${cast:str,'7'}")),
+            "age",
+            7,
+            IntegerNode,
+            id="convert_str_to_int",
+        ),
+        pytest.param(
+            MissingList(list=SI("${identity:[a, b, c]}")),
+            "list",
+            ["a", "b", "c"],
+            ListConfig,
+            id="list_str",
+        ),
+        pytest.param(
+            MissingList(list=SI("${identity:[0, 1, 2]}")),
+            "list",
+            ["0", "1", "2"],
+            ListConfig,
+            id="list_int_to_str",
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{key1: val1, key2: val2}}")),
+            "dict",
+            {"key1": "val1", "key2": "val2"},
+            DictConfig,
+            id="dict_str",
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{a: 0, b: 1}}")),
+            "dict",
+            {"a": "0", "b": "1"},
+            DictConfig,
+            id="dict_int_to_str",
+        ),
+    ],
+)
+def test_interpolation_type_validated_ok(
+    cfg: Any,
+    key: str,
+    expected_value: Any,
+    expected_node_type: Any,
+    common_resolvers: Any,
+) -> Any:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
+
+    def drop_last(s: str) -> str:
+        return s[0:-1]  # drop last character from string `s`
+
+    OmegaConf.register_new_resolver("drop_last", drop_last)
+
+    cfg = OmegaConf.structured(cfg)
+
+    val = cfg[key]
+    assert val == expected_value
+
+    node = cfg._get_node(key)
+    assert isinstance(node, Node)
+    assert isinstance(node._dereference_node(), expected_node_type)
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key", "expected_error"),
+    [
+        pytest.param(
+            User(name="Bond", age=SI("${cast:str,seven}")),
+            "age",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape(
+                    dedent(
+                        """\
+                        Value 'seven' could not be converted to Integer
+                            full_key: age
+                        """
+                    )
+                ),
+            ),
+            id="type_mismatch_resolver",
+        ),
+        pytest.param(
+            User(name="Bond", age=SI("${name}")),
+            "age",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape(
+                    dedent(
+                        """\
+                        Value 'Bond' could not be converted to Integer
+                            full_key: age
+                        """
+                    )
```

Comment:
I think for something like this we can omit testing the a the complete error. just go for the message.
```suggestion
                match=re.escape("'Bond' could not be converted to Integer")
```

Reply:
Taking suggestion => resolving